### ISSUE TITLE
Recipe for overfitting a single obj

### ISSUE THREAD

**Issue by ZekaiGalaxy** [lucidrains/meshgpt-pytorch#38]
Hi Happy New Year :) and sorry to ask again.

I failed to overfit a single obj (the autoencoder recon loss is always around 0.7).

Currently I am using sampled 100 tables from ShapeNetV2 as the dataset without any augmentation, and I follow the ipynb written by @MarcusLoppe. So can I kindly ask for the "data recipe" if I want to overfit a table as the first step.

Thanks a lot.

**Comment by MarcusLoppe** [comment]
> Hi Happy New Year :) and sorry to ask again.
> 
> I failed to overfit a single obj (the autoencoder recon loss is always around 0.7).
> 
> Currently I am using sampled 100 tables from ShapeNetV2 as the dataset without any augmentation, and I follow the ipynb written by @MarcusLoppe. So can I kindly ask for the "data recipe" if I want to overfit a table as the first step.
> 
> Thanks a lot.

When you say 100 tables, is this different tables or the same?
To be able to generalize that will requires I think a lot more data and training time.

If you just want to train a model for a single table, try the below. Autoencoder should be at 0.3 and the transformer at 0.001 - 0.0001.

p.s i just smacked together the code, might be some easy fixable mistake.
```
def get_mesh(file_path): 
    mesh = trimesh.load(file_path, force='mesh')
     
    vertices = mesh.vertices.tolist()
    faces = mesh.faces.tolist()
    centered_vertices = vertices - np.mean(vertices, axis=0)
 
    max_abs = np.max(np.abs(centered_vertices))
    vertices = centered_vertices / (max_abs / 0.95)   
      
    def sort_vertices(vertex):
        return vertex[1], vertex[0], vertex[2]   
 
    seen = OrderedDict()
    for point in vertices: 
      key = tuple(point)
      if key not in seen:
        seen[key] = point
        
    unique_vertices =  list(seen.values()) 
    sorted_vertices = sorted(unique_vertices, key=sort_vertices)
     
    vertices_as_tuples = [tuple(v) for v in vertices]
    sorted_vertices_as_tuples = [tuple(v) for v in sorted_vertices]

    vertex_map = {old_index: new_index for old_index, vertex_tuple in enumerate(vertices_as_tuples) for new_index, sorted_vertex_tuple in enumerate(sorted_vertices_as_tuples) if vertex_tuple == sorted_vertex_tuple}
 

    reindexed_faces = [[vertex_map[face[0]], vertex_map[face[1]], vertex_map[face[2]]] for face in faces] 
    return np.array(sorted_vertices), np.array(reindexed_faces)
```

MeshDataset (https://github.com/MarcusLoppe/meshgpt-pytorch/commit/9766d41999c0dccdeb15d45a123f73710d4c51ad)


```
vertices , faces = get_mesh("table.obj")
obj_data = {"vertices": torch.tensor(vertices.tolist(), dtype=torch.float).to("cuda"), "faces":  torch.tensor(faces.tolist(), dtype=torch.long).to("cuda"),"texts": "table"}

dataset = MeshDataset ( [obj_data] )
dataset.generate_face_edges()
dataset.data = [dataset[0]] * 500

autoencoder_trainer = MeshAutoencoderTrainer(model =autoencoder,learning_rate = 1e-1, 
                                             warmup_steps = 10,
                                             dataset = dataset, 
                                             checkpoint_every_epoch = 20,  
                                             num_train_steps=100,
                                             batch_size=16,
                                             grad_accum_every=1) 
loss = autoencoder_trainer.train(30 ,stop_at_loss = 0.25)   

max_length =  max(len(d["faces"]) for d in dataset if "faces" in d) 
max_seq =  max_length * 6   
transformer = MeshTransformer(
    autoencoder,
    dim = 512,
    coarse_pre_gateloop_depth = 6,
    fine_pre_gateloop_depth= 4, 
    max_seq_len = max_seq, 
)
 
dataset.generate_codes(autoencoder,1)

trainer = MeshTransformerTrainer(model = transformer,warmup_steps = 10,grad_accum_every=1,num_train_steps=100,   dataset = dataset, learning_rate = 1e-1, batch_size=16 , checkpoint_every_epoch=5)
loss = trainer.train(50,stop_at_loss = 0.00009) 
```


Display it with different temps.
```
coords = []
    for r in np.arange(0, 1.0, 0.1):
        faces_coordinates = transformer.generate(temperature=r ) 
        coords.append(faces_coordinates) 
    
    all_vertices = []
    all_faces = []
    vertex_offset = 0

    # Translation distance for each model
    translation_distance = 0.3  # Adjust as needed

    for r, faces_coordinates in enumerate(coords): 
        tensor_data = faces_coordinates[0].cpu()

        numpy_data = tensor_data.numpy().reshape(-1, 3)

        # Translate the model to avoid overlapping
        numpy_data[:, 0] += translation_distance * (r / 0.2 - 1)  # Adjust X coordinate

        # Accumulate vertices
        for vertex in numpy_data:
            all_vertices.append(f"v {vertex[0]} {vertex[1]} {vertex[2]}\n")

        # Accumulate faces with adjusted indices
        for i in range(1, len(numpy_data), 3):
            all_faces.append(f"f {i + vertex_offset} {i + 1 + vertex_offset} {i + 2 + vertex_offset}\n")

        # Update the vertex offset for the next model
        vertex_offset += len(numpy_data)

    # Combine vertices and faces
    obj_file_content = "".join(all_vertices) + "".join(all_faces)

    # Save to a single file
    obj_file_path = f"./results/3d_models_{text}_temps.obj"
    with open(obj_file_path, "w") as file:
        file.write(obj_file_content)
```

**Comment by ZekaiGalaxy** [comment]
I trained with 100 different chairs. Thank you so much, I will try it out!

**Comment by ZekaiGalaxy** [comment]
<img width="380" alt="image" src="https://github.com/lucidrains/meshgpt-pytorch/assets/77915605/64a1fc35-410c-4838-bfa5-944f584c4158">
HahaHaha It works, move on to generalize to more objects!

**Comment by ZekaiGalaxy** [comment]
<img width="380" alt="image" src="https://github.com/lucidrains/meshgpt-pytorch/assets/77915605/41b9f55c-3937-4f67-b607-7110983d5cf9">

Haha It works, move on to generalize to more objs!

**Comment by ZekaiGalaxy** [comment]
Generalize to 5 different chairs! But it's a torture to adjust the hyperparams again and again :(
<img width="947" alt="image" src="https://github.com/lucidrains/meshgpt-pytorch/assets/77915605/1a246cd0-bb6b-4dc9-863f-64fbaaa7d23d">

**Comment by wander2017** [comment]
> Generalize to 5 different chairs! But it's a torture to adjust the hyperparams again and again :( <img alt="image" width="947" src="https://private-user-images.githubusercontent.com/77915605/292976760-1a246cd0-bb6b-4dc9-863f-64fbaaa7d23d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDM3NTQ4NDcsIm5iZiI6MTcwMzc1NDU0NywicGF0aCI6Ii83NzkxNTYwNS8yOTI5NzY3NjAtMWEyNDZjZDAtYmI2Yi00ZGM5LTg2M2YtNjRmYmFhYTdkMjNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMjI4VDA5MDkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTYyZjRkZTgyZWQ5ZTQ4NDdhOGIwYmI0ZWViNTczYWE0MjNmNjg4OGY1N2IxMzFjYWI4M2VkNWNiZTc0MjVjNWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.iU3mlMOkywB_vLU9smrYppeGuvdh46gzZdMxiEUuE-o">

请问tranfomer的loss收敛到多少才有这种效果啊？我用4个模型一起训练，重建只有一个形状成功了，其他看上去没有收敛

**Comment by lucidrains** [comment]
> Generalize to 5 different chairs! But it's a torture to adjust the hyperparams again and again :( <img alt="image" width="947" src="https://private-user-images.githubusercontent.com/77915605/292976760-1a246cd0-bb6b-4dc9-863f-64fbaaa7d23d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ5MDI2NDcsIm5iZiI6MTcwNDkwMjM0NywicGF0aCI6Ii83NzkxNTYwNS8yOTI5NzY3NjAtMWEyNDZjZDAtYmI2Yi00ZGM5LTg2M2YtNjRmYmFhYTdkMjNkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMTAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTEwVDE1NTkwN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4YjM4MDdkODZmZTgxNGEwZWM0MTk3MTE5MzQzNDdiYWViZmFhMzAwNjg3ZDNlMDZjNzZmZDFiYzNmM2E2ZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-OL5MO1n5hgmnMOAAVA_iExISd6QtkuomlOZ1KEicHM">

nice! good starting point
